### DISCUSSION TITLE
Pattern matching is limited by overlapping patterns

### DISCUSSION THREAD

**Discussion by refparo** [aya-prover/aya-dev#98] [General]
Consider the following code, which is essentially almost identical to the code in [#96](https://github.com/aya-prover/aya-dev/discussions/96#discussioncomment-1500198), only with the two clauses swapped.

```aya
open data Bool : Type 0 | true | false
def not (b : Bool) : Bool
 | false => true
 | otherwise => false
```

The second clause is intended as a "default" case. Such pattern may be widely used in real-world programming, but Aya (as of 1747794) rejects this. Is this behavior expected? Is it possible to make this type check while also having overlapping patterns?

**Comment by ice1000** [comment]
It is rejected. The solution would be to let the two styles of function definitions coexist, but we need to work on that. It's not like this is impossible.

**Reply by ice1000** [reply]
See also #190

**Comment by ice1000** [comment]
We now support these definitions by default. Overlapping patterns need explicit specification `overlap`
